test-network-function/networking/suite.go

```diff
@@ -324,3 +348,99 @@ func testNodePort(env *config.TestEnvironment) {
+	ocCommandToExecute := fmt.Sprintf(commandportdeclared, podName, podNamespace, container)
+	res, _ := utils.ExecuteCommand(ocCommandToExecute, ocCommandTimeOut, interactive.GetContext(false))
+	err := parseVariables(res, declaredPorts)
+	return err
+}
+
+func listeningPortList(commandlisten []string, nodeOc *interactive.Context, listeningPorts map[key]string) {
+	var k key
+	listeningPortCommand := strings.Join(commandlisten, " ")
+	fmt.Println(listeningPortCommand)
+	res, _ := utils.ExecuteCommand(listeningPortCommand, ocCommandTimeOut, nodeOc)
+	if res == "" {
+		return
+	}
+	lines := strings.Split(res, "\n")
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if indexprotocolname > len(fields) || indexport > len(fields) {
+			return
+		}
+		s := strings.Split(fields[indexport], ":")
+		p, _ := strconv.Atoi(s[1])
+		k.port = p
+		k.protocol = strings.ToUpper(fields[indexprotocolname])
+		listeningPorts[k] = ""
+	}
+}
+
+func checkIfListenIsDeclared(listeningPorts, declaredPorts map[key]string) bool {
+	if len(listeningPorts) == 0 || len(declaredPorts) == 0 {
+		return false
+	}
+	for k := range listeningPorts {
+		_, ok := declaredPorts[k]
+		if !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func testListenAndDeclared(env *config.TestEnvironment) {
+	declaredPorts := make(map[key]string)
+	listeningPorts := make(map[key]string)
+	var x *configsections.Container
+	testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestServicesDoNotUseNodeportsIdentifier)
+	ginkgo.It(testID, func() {
+		for _, podUnderTest := range env.PodsUnderTest {
+			for i := 0; i < podUnderTest.ContainerCount; i++ {
+				err := declaredPortList(i, podUnderTest.Name, podUnderTest.Namespace, declaredPorts)
+				if err != nil {
+					return
+				}
+			}
+
+			nodeName := podnodename.NewPodNodeName(common.DefaultTimeout, podUnderTest.Name, podUnderTest.Namespace)
+			context := env.GetLocalShellContext()
+			test, err := tnf.NewTest(context.GetExpecter(), nodeName, []reel.Handler{nodeName}, context.GetErrorChannel())
+			gomega.Expect(err).To(gomega.BeNil())
+			test.RunAndValidate()
+			nodeOc := env.NodesUnderTest[nodeName.GetNodeName()].DebugContainer.GetOc()
+			for _, cut := range env.ContainersUnderTest {
+				if cut.ContainerIdentifier.PodName == podUnderTest.Name && cut.ContainerIdentifier.Namespace == podUnderTest.Namespace {
+					x = cut
+					break
+				}
+			}
+			containerPID := utils.GetContainerPID(nodeName.GetNodeName(), nodeOc, x.ContainerUID, x.ContainerRuntime)
+
+			commandlisten := []string{utils.AddNsenterPrefix(containerPID), commandportlisten}
+
+			listeningPortList(commandlisten, nodeOc.Context, listeningPorts)
+			// compare between declaredPort,listeningPort and return the common.
+			res := checkIfListenIsDeclared(listeningPorts, declaredPorts)
+			if !res {
+				ginkgo.Fail("TC failed : port is listening but not declared.")
```

Comment:
I meant actually reporting which listening ports were not declared

Reply:
You can use the ginkgo log writer function to dump the bad ports while finding it. It's probably easier than collecting everything and dumping them at the end.

Reply:
Also, we don't call fail on the first failure. Need to finish testing other pods

Reply:
Fixed, I was replace the "gingko.Fail" to "log.info", if you prefer something else tell me. I was search gingko log but I didn't find.